lib/Doctrine/DBAL/DriverManager.php

```diff
@@ -115,11 +115,15 @@ private function __construct()
      * <b>driverClass</b>:
      * The driver class to use.
      *
-     * @param mixed[]            $params       The parameters.
-     * @param Configuration|null $config       The configuration to use.
-     * @param EventManager|null  $eventManager The event manager to use.
+     * @param array{wrapperClass?: class-string<T>} $params       The parameters.
+     * @param Configuration|null                    $config       The configuration to use.
+     * @param EventManager|null                     $eventManager The event manager to use.
+     *
+     * @return T
```

Comment:
I'm going to revert this, because if you pass in `null` as a `wrapperClass`, then it returns a `Connection`, and I don't think there is way to express that it return `Connection` if null, or `T` otherwise.

Reply:
Would it help if we didn't ignore `NULL and used `array_key_exists()` instead of `isset()`. API-wise, what's the point in passing an option with `NULL` and expecting that it will be ignored?

Reply:
I just tried, and it does not help Psalm. These errors come back as soon as I add the `|Connection`, and using `array_key_exists` does not improve that situation:

```
ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:193:33 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertNull($conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:196:38 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertEquals(0, $conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:199:38 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertEquals(1, $conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:202:33 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertNull($conn->getActiveShardId());
```

Reply:
I just asked for support here, just in case: https://symfony-devs.slack.com/archives/C8SFXTD2M/p1588144518359100

Reply:
This is going to get fixed: https://github.com/vimeo/psalm/issues/3277 :)

Reply:
I have it working locally, but now PHPStan is unhappy about it :see_no_evil:

Reply:
I pushed a new commit that include additional rules to ignor ethe PHPStan issues this creates.